src/team_lib/mhba/xccl_mhba_mkeys.c

```diff
@@ -0,0 +1,283 @@
+    int total_received_completions = 0;
+    int new_received_completions = 0;
+    int completions_to_wait_for = expected_wc_num;
+
+    // TODO: Consider doing this instead of requiring the user to initialize
+    // the array of completions.
+    //memset((void*)actual_wc, 0, expected_wc_num*sizeof(struct ibv_wc));
+
+    while ((completions_to_wait_for > 0) && (poll_for_completion_retry > 0)) {
+        new_received_completions = ibv_poll_cq(cq, completions_to_wait_for, &actual_wc[total_received_completions]);
+        if (new_received_completions < 0) {
+            xccl_mhba_error("ibv_poll_cq() failed");
+            return new_received_completions;
+        }
+        total_received_completions += new_received_completions;
+        completions_to_wait_for -= new_received_completions;
+        if (completions_to_wait_for > 0) {
+            poll_for_completion_retry--;
+            sleep(1); // todo ?
+        }
+    }
+
+    if (total_received_completions == 0) {
+        xccl_mhba_error("No CQEs were generated. Retry counter exceeded");
+        return -1;
+    }
+
+    if (completions_to_wait_for > 0) {
+        xccl_mhba_error("Not enough CQE were generated. Retry counter exceeded");
+        return total_received_completions;
+    }
+    *actual_wc_num = total_received_completions;
+    return total_received_completions;
+}
+
+xccl_status_t create_umr_qp(xccl_mhba_context_t *ctx) {
+    struct ibv_qp_init_attr_ex umr_init_attr_ex;
+    struct mlx5dv_qp_init_attr umr_mlx5dv_qp_attr;
+
+    xccl_mhba_debug("Create UMR QP (and CQ)");
+    ctx->umr_cq = ibv_create_cq(ctx->ib_ctx, UMR_CQ_SIZE, NULL, NULL, 0);
+    if (ctx->umr_cq == NULL) {
+        xccl_mhba_error("UMR CQ creation failed");
+        goto umr_cq_creation_failed;
+    }
+
+    memset(&umr_mlx5dv_qp_attr, 0, sizeof(umr_mlx5dv_qp_attr));
+    umr_mlx5dv_qp_attr.comp_mask |= MLX5DV_QP_INIT_ATTR_MASK_SEND_OPS_FLAGS, IBV_QP_INIT_ATTR_PD;
+    umr_mlx5dv_qp_attr.create_flags = 0;
+    memset(&umr_mlx5dv_qp_attr.dc_init_attr,0,sizeof(umr_mlx5dv_qp_attr.dc_init_attr)); //todo: check is right
+    umr_mlx5dv_qp_attr.send_ops_flags = MLX5DV_QP_EX_WITH_MR_LIST | MLX5DV_QP_EX_WITH_MR_INTERLEAVED;
+
+    memset(&umr_init_attr_ex, 0, sizeof(umr_init_attr_ex));
+    umr_init_attr_ex.send_cq = ctx->umr_cq;
+    umr_init_attr_ex.recv_cq = ctx->umr_cq;
+    umr_init_attr_ex.cap.max_send_wr = 1;
+    umr_init_attr_ex.cap.max_recv_wr = 0;
+    umr_init_attr_ex.cap.max_send_sge = 1;
+    umr_init_attr_ex.cap.max_recv_sge = 0;
+    // `max_inline_data` determines the WQE size that the QP will support.
+    // The 'max_inline_data' should be modified only when the number of
+    // arrays to interleave is greater than 3.
+    umr_init_attr_ex.cap.max_inline_data = 256; // todo: in final stage the UMR won't be a WQE but a raw request, in this stage we might hit a ceiling with that size, need to test-and-err. for big sbgp_size might be a problem, depend on number of UMR entries
```

Comment:
comment line too long.
second: i don't understand the limitation here. What is actually limited to 256 bytes? Why do we have to use INLINE to create umr?

Reply:
@vspetrov the limitation arrive from the NIC - limited memory to contain the actual WQE request. as we have more UMR elements, we will need more memory. this field will determine how much memory we allocate, but it also have a limit.
In final stage, we won't use this API for UMR, but some sort of raw API. in this case, NIC will be able to use secondary memory for the WQE if needed, and this limitation will be gone.
for now, I asked Rami to create a calculation(#entries->needed max_inline_data), plus understanding what s the max memory(to know max block size possible), but since it won't be needed for later stages, we agreed to just keep it as an open issue for now, and remember to play with this field as we actually try the algorithm.

Reply:
@lappazos  So, sizeof(struct mlx5dv_mr_interleaved) = 20. Does this mean that we can only support at most 12 ranks per node right now ? I mean, do i understand the limit correctly - is this the size of mr_interleaved array?